File: mpp-proiect-repository-cristinahendre/lab6(web)/Java/lab6/modul/RepoModule/src/main/java/Repository/DonatieRepo.java
package Repository;

import Domain.CazCaritabil;
import Domain.Donatie;
import Domain.Donator;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class DonatieRepo implements  DonatieRepoI {

    private CazRepoI cazRepo;
    private DonatorRepoI donatorRepo;

    private JdbcUtils utils;

    public DonatieRepo(Properties prop, CazRepoI cazRepo, DonatorRepoI donatorRepo) {
        logger.info("intializare cu proprietati {}", prop);
        utils= new JdbcUtils(prop);
        this.cazRepo=cazRepo;
        this.donatorRepo=donatorRepo;
    }

    private  static  final Logger logger = LogManager.getLogger();


    @Override
    public Donatie findOne(Integer integer) {
        logger.traceEntry("Caut donatia cu acest id: {}", integer);

        Connection con = utils.getConnection();

        try(PreparedStatement statement = con.prepareStatement("select * from donatie where id = ?")){

            statement.setInt(1,integer);
            try(ResultSet resultSet = statement.executeQuery()){
                if(resultSet.next()){
                    int id = resultSet.getInt(1);
                    int caz_id = resultSet.getInt(2);
                    int donator_id = resultSet.getInt(3);
                    int suma =resultSet.getInt(4);
                    CazCaritabil caz= cazRepo.findOne(caz_id);
                    Donator donator = donatorRepo.findOne(donator_id);
                    Donatie d = new Donatie(donator,caz,suma);
                    d.setId(id);
                    return  d;

                }
            }

            con.close();

        } catch (SQLException throwables) {
            throwables.printStackTrace();
        }
        return null;

    }

    @Override
    public Iterable<Donatie> findAll() {
        logger.traceEntry("afisez toate donatiile ");
        List<Donatie> lista = new ArrayList<>();
        Connection con = utils.getConnection();

        try(PreparedStatement pst = con.prepareStatement("select * from donatie")){

            try(ResultSet resultSet = pst.executeQuery()){
                while(resultSet.next()){
                    int id = resultSet.getInt("id");
                    int caz_id= resultSet.getInt(2);
                    int donator_id =resultSet.getInt(3);
                    int suma =resultSet.getInt("suma");

                    CazCaritabil caz= cazRepo.findOne(caz_id);
                    Donator donator = donatorRepo.findOne(donator_id);
                    Donatie d = new Donatie(donator,caz,suma);
                    d.setId(id);
                    lista.add(d);


                }
            }
            con.close();
        } catch (SQLException throwables) {
            throwables.printStackTrace();
        }

        return  lista;
    }


    @Override
    public Donatie save(Donatie entity) {
        logger.traceEntry("Adaug donatia {}", entity);

        CazCaritabil caz = entity.getCaz();
        caz.setSuma_donata(caz.getSuma_donata() + entity.getSuma_donata());
        cazRepo.update(entity.getCaz());
        Connection con = utils.getConnection();


        try(PreparedStatement statement = con.prepareStatement
                ("insert into donatie(id_caz,id_donator,suma) values (?,?,?)")){


            statement.setInt(1,entity.getCaz().getId());
            statement.setInt(2,entity.getDonator().getId());
            statement.setInt(3,entity.getSuma_donata());

            int result= statement.executeUpdate();
            logger.traceEntry("S-au modificat {} randuri",result);
            //con.close();
            return entity;



        } catch (SQLException throwables) {
            throwables.printStackTrace();
        }
        return  null;
    }

    @Override
    public void delete(Integer integer) {

        Donatie dn = findOne(integer);
        CazCaritabil caz = dn.getCaz();
        caz.setSuma_donata(caz.getSuma_donata()-dn.getSuma_donata());
        cazRepo.update(caz);


        logger.traceEntry("Sterg donatia cu id-ul {}", integer);
        Connection con = utils.getConnection();

        try(PreparedStatement statement = con.prepareStatement
                ("delete from donatie where id = ?")){

            statement.setInt(1,integer);

            int result= statement.executeUpdate();
            logger.traceEntry("S-au modificat {} randuri",result);
            con.close();




        } catch (SQLException throwables) {
            throwables.printStackTrace();
        }


    }




    @Override
    public Donatie update(Donatie entity) {
        logger.traceEntry("Modific donatia {}", entity);
        Donatie initial = findOne(entity.getId());
        if(initial.getCaz().equals(entity.getCaz())){
            //se modifica doar suma donata
            CazCaritabil caz= initial.getCaz();
            caz.setSuma_donata(caz.getSuma_donata() - initial.getSuma_donata() + entity.getSuma_donata());

            cazRepo.update(caz);
        }
        else{

            //se modifica acel caz
            CazCaritabil caz= initial.getCaz();
            caz.setSuma_donata(caz.getSuma_donata()- initial.getSuma_donata());
            cazRepo.update(caz);

            CazCaritabil nou =entity.getCaz();
            nou.setSuma_donata(nou.getSuma_donata() + entity.getSuma_donata());
            cazRepo.update(nou);

        }


        Connection con = utils.getConnection();

        try(PreparedStatement statement = con.prepareStatement
                ("update donatie set id_caz= ? , id_donator=?, suma =? where id =?")){

            statement.setInt(1,entity.getCaz().getId());
            statement.setInt(2,entity.getDonator().getId());
            statement.setInt(3,entity.getSuma_donata());
            statement.setInt(4,entity.getId());

            int result= statement.executeUpdate();
            logger.traceEntry("S-au modificat {} randuri",result);
            con.close();
            return entity;



        } catch (SQLException throwables) {
            throwables.printStackTrace();
        }
        return  null;
    }

    @Override
    public int nrElem() {
        logger.traceEntry("Numar cate elemente sunt in donatie");
        int nr =0;
        Connection con = utils.getConnection();
        try(Statement statement = con.createStatement()){

            ResultSet resultSet = statement.executeQuery("select count(*) from donatie");
            resultSet.next();
            nr= resultSet.getInt(1);
            con.close();
            return nr;


        } catch (SQLException throwables) {
            throwables.printStackTrace();
        }
        logger.traceExit(nr);
        return nr;
    }

    @Override
    public Donatie getDonatieDupaDate(Donatie don) {
        logger.traceEntry("Caut donatia cu aceste date: {}", don);

        Connection con = utils.getConnection();

        try(PreparedStatement statement = con.prepareStatement("select id from donatie where id_caz = ? and id_donator =? and suma =?")){

            statement.setInt(1,don.getCaz().getId());
            statement.setInt(2,don.getDonator().getId());
            statement.setInt(3,don.getSuma_donata());

            try(ResultSet resultSet = statement.executeQuery()){
                if(resultSet.next()){
                    int id = resultSet.getInt(1);


                    don.setId(id);
                    return  don;

                }
            }

           // con.close();

        } catch (SQLException throwables) {
            throwables.printStackTrace();
        }
        return null;

    }
}

File: mpp-proiect-repository-cristinahendre/lab6(web)/CSharp/Lab5/Repository/DonatieRepo.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

using Cerinta1.Domain;
using Cerinta1.Domain.Validator;
using log4net;

namespace Cerinta1.Repository
{
    public class DonatieRepo : IDonatieRepo
    {

        ICazRepo cazRepo;
        IDonatorRepo donatorRepo;
        private static readonly ILog log = LogManager.GetLogger("DonatieRepo");

        public DonatieRepo(ICazRepo cazRepo, IDonatorRepo donatorRepo)
        {
            this.cazRepo = cazRepo;
            this.donatorRepo = donatorRepo;
        }

        public Donatie Delete(int id)
        {
            Donatie dn = FindOne(id);
            CazCaritabil caz = dn.Caz;

            caz.Suma_donata = caz.Suma_donata - dn.Suma_donata;
            cazRepo.Update(caz);

            IDbConnection con = DBUtils.getConnection();
            log.InfoFormat("incerc sa sterg donatia {0}", id);
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "delete from donatie where id=@id";
                IDbDataParameter paramId = comm.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;
                comm.Parameters.Add(paramId);
                var dataR = comm.ExecuteNonQuery();
                con.Close();
                if (dataR == 0)
                {
                    log.InfoFormat("Eroare la stergerea donatiei.");
                    throw new RepositoryException("Nu s-a sters cu succes!");
                }
                else
                {
                    log.InfoFormat("Stergerea donatiei a avut loc cu succes.");
                    return null;
                }
            }
        }


        public IEnumerable<Donatie> FindAll()
        {
            IList<Donatie> lista = new List<Donatie>();
            log.InfoFormat("Afisez toate donatiile");
            IDbConnection con = DBUtils.getConnection();
            try
            {

                using (var comm = con.CreateCommand())
                {
                    comm.CommandText = "select * from donatie ";


                    using (var dataR = comm.ExecuteReader())
                    {
                        while (dataR.Read())
                        {
                            int id = dataR.GetInt32(0);
                            int id_caz = dataR.GetInt32(1);
                            int id_donator = dataR.GetInt32(2);
                            int suma = dataR.GetInt32(3);

                            CazCaritabil caz = cazRepo.FindOne(id_caz);
                            Donator donator = donatorRepo.FindOne(id_donator);
                            Donatie donatie = new Donatie(caz, suma, donator);
                            donatie.Id = id;

                            lista.Add(donatie);
                        }
                    }
                }
            }
            catch (Exception )
            {
                // MessageBox.Show(ex.Message);
            }
            con.Close();
            log.InfoFormat("Am parcurs donatiile.");
            return lista;
        }

        public Donatie FindOne(int id)
        {
            log.InfoFormat("Caut donatia cu acest id: {0}", id);
            IDbConnection con = DBUtils.getConnection();
            try
            {

                using (var comm = con.CreateCommand())
                {
                    comm.CommandText = "select * from donatie where id=@id";
                    IDbDataParameter paramId = comm.CreateParameter();
                    paramId.ParameterName = "@id";
                    paramId.Value = id;
                    comm.Parameters.Add(paramId);

                    using (var dataR = comm.ExecuteReader())
                    {
                        if (dataR.Read())
                        {
                            
                            int id_caz = dataR.GetInt32(1);
                            int id_donator = dataR.GetInt32(2);
                            int suma = dataR.GetInt32(3);

                            CazCaritabil caz = cazRepo.FindOne(id_caz);
                            Donator donator = donatorRepo.FindOne(id_donator);
                            Donatie donatie = new Donatie(caz, suma, donator);
                            donatie.Id = id;
                            con.Close();
                            return donatie;
                        }
                    }
                }
            }
            catch (Exception )
            {
                //  MessageBox.Show(ex.Message);
            }
            log.InfoFormat("nu am  gasit acea donatie.");
            return null;

        }

        public int nrElem()
        {
            log.InfoFormat("Numar cate donatii sunt in baza de date.");
            IDbConnection con = DBUtils.getConnection();
            try
            {
                using (var comm = con.CreateCommand())
                {
                    comm.CommandText = "select count(*) from donatie";
                    int rez = int.Parse(comm.ExecuteScalar().ToString());
                    con.Close();
                    log.InfoFormat("Sunt {0} donatii.", rez);
                    return rez;
                }


            }
            catch (Exception )
            {
                //  MessageBox.Show(ex.Message);
            }
            return -1;
        }

        public Donatie Save(Donatie entity)
        {
            log.InfoFormat("Voi salva donatia {0}", entity);
            CazCaritabil caz = entity.Caz;
            caz.Suma_donata = caz.Suma_donata + entity.Suma_donata;
            cazRepo.Update(entity.Caz);
            IDbConnection con = DBUtils.getConnection();


            try
            {
               
                    using (var comm = con.CreateCommand())
                    {
                        comm.CommandText = "insert into donatie(id_caz,id_donator,suma) values (@caz,@don,@s)";



                        IDbDataParameter paramCaz = comm.CreateParameter();
                        paramCaz.ParameterName = "@caz";
                        paramCaz.Value = entity.Caz.Id;
                        comm.Parameters.Add(paramCaz);

                        IDbDataParameter paramDonator = comm.CreateParameter();
                        paramDonator.ParameterName = "@don";
                        paramDonator.Value = entity.Donator.Id;


                        comm.Parameters.Add(paramDonator);

                        IDbDataParameter paramSuma = comm.CreateParameter();
                        paramSuma.ParameterName = "@s";
                        paramSuma.Value = entity.Suma_donata;


                        comm.Parameters.Add(paramSuma);

                        var dataR = comm.ExecuteNonQuery();

                        if (dataR == 0)
                        {
                            log.InfoFormat("Salvarea donatiei a intampinat probleme");
                            throw new RepositoryException("Nu s-a adaugat!");
                        }
                        else
                        {
                            log.InfoFormat("Salvarea donatiei e cu succes.");

                            return entity;
                        }
                    }
                }
            
            catch (Exception )
            {
                //  MessageBox.Show(ex.Message);
            }
            finally
            {
             //  con.Close();
            }
            return null;
        }



        public Donatie Update(Donatie entity)
        {
            Donatie initial = FindOne(entity.Id);

            if (initial.Caz.Equals(entity.Caz))
            {
                //se modifica doar suma donata
                CazCaritabil caz = initial.Caz;
                caz.Suma_donata = caz.Suma_donata - initial.Suma_donata + entity.Suma_donata;

                cazRepo.Update(caz);
            }
            else
            {

                //se modifica acel caz
                CazCaritabil caz = initial.Caz;
                caz.Suma_donata = caz.Suma_donata - initial.Suma_donata;
                cazRepo.Update(caz);

                CazCaritabil nou = entity.Caz;
                nou.Suma_donata = nou.Suma_donata + entity.Suma_donata;
                cazRepo.Update(nou);

            }



            log.InfoFormat("Modific donatia {0}", entity);
            try
            {
                IDbConnection con = DBUtils.getConnection();
                using (var comm = con.CreateCommand())
                {
                    comm.CommandText = "update donatie set id_caz=@caz, id_donator =@don ,suma=@s where id = @id";
                    IDbDataParameter paramId = comm.CreateParameter();
                    paramId.ParameterName = "@id";
                    paramId.Value = entity.Id;



                    IDbDataParameter paramCaz = comm.CreateParameter();
                    paramCaz.ParameterName = "@caz";
                    paramCaz.Value = entity.Caz.Id;
                    comm.Parameters.Add(paramCaz);

                    IDbDataParameter paramDonator = comm.CreateParameter();
                    paramDonator.ParameterName = "@don";
                    paramDonator.Value = entity.Donator.Id;
                    comm.Parameters.Add(paramDonator);


                    IDbDataParameter paramSuma = comm.CreateParameter();
                    paramSuma.ParameterName = "@s";
                    paramSuma.Value = entity.Suma_donata;


                    comm.Parameters.Add(paramSuma);
                    comm.Parameters.Add(paramId);

                    var dataR = comm.ExecuteNonQuery();
                    con.Close();
                    if (dataR == 0)
                    {
                        log.InfoFormat("Eroare la modificarea donatiei.");
                        throw new RepositoryException("Nu s-a modificat!");
                    }
                    else
                    {
                        log.InfoFormat("Donatie modificata cu succes.");
                        return entity;
                    }
                }
            }
            catch (Exception )
            {
                //  MessageBox.Show(ex.Message);
            }
            return null;
        }
    }
}
